File: src/one_r.rs
use std::collections::HashMap;

pub struct OneRClassifier {
    best_feature: usize,
    rules: HashMap<String, String>,
    default_class: String,
}

impl OneRClassifier {
    pub fn new() -> Self {
        OneRClassifier {
            best_feature: 0,
            rules: HashMap::new(),
            default_class: String::new(),
        }
    }

    pub fn fit(&mut self, x: &Vec<Vec<String>>, y: &Vec<String>) {
        let num_features = x[0].len();
        let mut best_error = usize::MAX;
        let mut best_feature_index = 0;
        let mut best_rules: HashMap<String, String> = HashMap::new();

        let mut class_counts: HashMap<&String, usize> = HashMap::new();
        for label in y {
            *class_counts.entry(label).or_insert(0) += 1;
        }
        if let Some((majority_class, _)) = class_counts.iter().max_by_key(|entry| entry.1) {
            self.default_class = (*majority_class).clone();
        }

        for fi in 0..num_features {
            let mut value_class_counts: HashMap<String, HashMap<String, usize>> = HashMap::new();
            for (index, (row, label)) in x.iter().zip(y.iter()).enumerate() {
                let value = &row[fi];
                value_class_counts
                    .entry(value.clone())
                    .or_insert(HashMap::new())
                    .entry(label.clone())
                    .or_insert(0_usize);
                *value_class_counts.get_mut(value).unwrap().get_mut(label).unwrap() += 1;
                let count = value_class_counts.get(value).unwrap().get(label).unwrap();
            }

            let mut rules: HashMap<String, String> = HashMap::new();
            let mut errors = 0;
            for (value, class_map) in &value_class_counts {
                let majority_class = class_map
                    .iter()
                    .max_by_key(|entry| entry.1)
                    .map(|(class, _)| class.clone())
                    .unwrap();
                rules.insert(value.clone(), majority_class.clone());

                let total_for_value: usize = class_map.values().sum();
                let correct_count = class_map.get(&majority_class).unwrap();
                errors += total_for_value - *correct_count;
            }

            if errors < best_error {
                best_error = errors;
                best_feature_index = fi;
                best_rules = rules;
            }
        }

        self.best_feature = best_feature_index;
        self.rules = best_rules;
    }

    pub fn predict(&self, data: &[Vec<String>]) -> Vec<String> {
        let mut predictions: Vec<String> = Vec::new();
        for features in data {
            if let Some(value) = features.get(self.best_feature) {
                let predicted = self.rules.get(value)
                                   .unwrap_or(&self.default_class)
                                   .clone();
                predictions.push(predicted);
            } else {
                predictions.push(self.default_class.clone());
            }
        }
        predictions
    }

    pub fn get_best_feature_index(&self) -> usize {
        self.best_feature
    }
}

File: algorithms/one_r.py
import numpy as np
from sklearn.base import BaseEstimator, ClassifierMixin

class OneRClassifierCustom(BaseEstimator, ClassifierMixin):
    def __init__(self):
        self.best_feature = None        # Індекс найкращої ознаки після навчання
        self.rules = {}                 # Правила: значення ознаки -> передбачений клас
        self.default_class = None       # Клас за замовчуванням (якщо зустрінеться нове значення ознаки)

    def fit(self, X, y):
        X = np.asarray(X)
        y = np.asarray(y)
        
        num_features = X.shape[1]              # Кількість ознак (стовпців)
        best_error = float('inf')             # Мінімальна кількість помилок (кращий результат)
        best_rules = None
        best_feature_index = None
        
        # Розрахувати загальний найбільш частий клас у датасеті (для default_class)
        unique_classes, class_counts = np.unique(y, return_counts=True)
        overall_majority_class = unique_classes[np.argmax(class_counts)]
        self.default_class = overall_majority_class  # клас за замовчуванням (глобальна мода)
        
        # Перебір кожної ознаки, щоб знайти найкращу
        for fi in range(num_features):
            # Підрахунок частоти класів для кожного значення ознаки fi
            value_class_counts = {}  # словник: значення ознаки -> {клас -> кількість}
            
            # Отримуємо значення поточної ознаки
            feature_values = X[:, fi]
            
            # Для кожного унікального значення ознаки
            for value in np.unique(feature_values):
                hashable_value = value.item() if hasattr(value, 'item') else value
                
                # Знаходимо всі індекси де значення ознаки дорівнює поточному значенню
                indices = np.where(feature_values == value)[0]
                
                # Підраховуємо класи для цих індексів
                value_classes = y[indices]
                unique_value_classes, value_class_counts_arr = np.unique(value_classes, return_counts=True)
                
                # Зберігаємо підрахунок у словнику
                value_class_counts[hashable_value] = {
                    cls: count for cls, count in zip(unique_value_classes, value_class_counts_arr)
                }
            
            # Визначення правил для ознаки fi та підрахунок помилок
            rules = {}   # правила для поточної ознаки: значення -> прогнозований клас
            errors = 0   # кількість помилок при використанні цієї ознаки
            
            for value, class_count_map in value_class_counts.items():
                # Найбільш частий клас для даного значення ознаки (правило)
                best_class_for_value = max(class_count_map, key=class_count_map.get)
                rules[value] = best_class_for_value
                
                # Помилки: усі випадки цього значення, що не належать до best_class_for_value
                total_for_value = sum(class_count_map.values())
                errors_for_value = total_for_value - class_count_map[best_class_for_value]
                errors += errors_for_value
            
            # Перевірка, чи ця ознака краща (менше помилок)
            if errors < best_error:
                best_error = errors
                best_feature_index = fi
                best_rules = rules
        
        # Зберегти кращу ознаку та її правила
        self.best_feature = best_feature_index
        self.rules = best_rules
        
        return self
    
    def predict(self, X):
        X = np.asarray(X)
        
        # Перевіряємо розмірність вхідних даних
        if len(X.shape) == 1:
            # Якщо це один приклад (одновимірний масив)
            value = X[self.best_feature]
            # Перетворюємо значення в хешований тип, якщо потрібно
            hashable_value = value.item() if hasattr(value, 'item') else value
            return self.rules.get(hashable_value, self.default_class)
        else:
            # Якщо це набір прикладів (двовимірний масив)
            predictions = []
            for row in X:
                value = row[self.best_feature]
                # Перетворюємо значення в хешований тип, якщо потрібно
                hashable_value = value.item() if hasattr(value, 'item') else value
                predicted_class = self.rules.get(hashable_value, self.default_class)
                predictions.append(predicted_class)
            return np.array(predictions)